### ISSUE TITLE
usage of connection_attempts raises select.error

### ISSUE THREAD

**Issue by GayatriNittala** [pika/pika#717]
Hi All,

I started using pika recently. I'm using pika 0.10.0 with RabbitMQ 3.6.0 & Python 2.7. Th ebelow script raises a select.error

import time
import pika

_HOST, _PORT = "localhost", 5672
conn_credentials = pika.PlainCredentials('guest', 'guest')
conn_parameters = pika.ConnectionParameters(host=_HOST, port=_PORT, virtual_host = r"/", credentials=conn_credentials, connection_attempts=10, retry_delay=1)
connection = pika.BlockingConnection(conn_parameters)
connection_channel = connection.channel()
print connection_channel
time.sleep(30)
print connection.is_open
print connection_channel.is_open
time.sleep(5)
connection_channel.close()
connection.close()

When the connection_attempts parameter is used - as shown in the script above, when the RMQ server is down, it raise Select.error as shown below
```
Traceback (most recent call last):
  File "C:\Gayatri\learning\pyrabbit\cloud_connect\testRabbitMQ\test_service_ret
ry.py", line 11, in <module>
    connection = pika.BlockingConnection(conn_parameters)
  File "C:\Python27\lib\site-packages\pika-0.10.0-py2.7.egg\pika\adapters\blocki
ng_connection.py", line 339, in __init__
    self._process_io_for_connection_setup()
  File "C:\Python27\lib\site-packages\pika-0.10.0-py2.7.egg\pika\adapters\blocki
ng_connection.py", line 374, in _process_io_for_connection_setup
    self._open_error_result.is_ready)
  File "C:\Python27\lib\site-packages\pika-0.10.0-py2.7.egg\pika\adapters\blocki
ng_connection.py", line 410, in _flush_output
    self._impl.ioloop.poll()
  File "C:\Python27\lib\site-packages\pika-0.10.0-py2.7.egg\pika\adapters\select
_connection.py", line 400, in poll
    self.get_next_deadline())
select.error: (10022, 'An invalid argument was supplied')
```
Is my understanding on usage of connection_attempts parameter correct? or is there anything else I need to do to retry the connection to the server?

Thanks for the help in advance.

Regards,
Gayatri

**Comment by GayatriNittala** [comment]
Sorry If this should not be posted here under the issues section. Posting this again under rabbitmq users group.

**Comment by vitaly-krugl** [comment]
@GayatriNittala: It looks like you're running on Windows. Windows CI testing is lacking in the pika project. It would be nice to get something like AppVeyor integrated into the pika project.

To debug this, configure logging to output (e.g., `import logging; logging.basicConfig(level=logging.DEBUG)`) and modify the section of code that calls select.select in SelectConnection like the following. The log output should help you to figure out which invalid argument is being passed to select. Please post here what you find out.

```
        while True:
            try:
                 read_fds = self._fd_events[READ]
                 write_fds = self._fd_events[WRITE]
                 error_fds = self._fd_events[ERROR]
                 next_deadline = self.get_next_deadline()

                 LOGGER.info('Calling select.select(%r, %r, %r, %r)', 
                             read_fds,
                             write_fds,
                             error_fds,
                             next_deadline)

                 read, write, error = select.select(read_fds,
                                                    write_fds,
                                                    error_fds,
                                                    next_deadline)


```

**Comment by GayatriNittala** [comment]
Thanks a lot for the quick response.

I shall add logging as suggested, but I did print the arguments passed to select.select

print self._fd_events[READ], self._fd_events[WRITE], self._fd_events[ERROR],self.get_next_deadline()
gives :
**set([]) set([]) set([]) 2.99799990654**

Python documentation states that, on Windows, select call raises an error if all the first arguments were empty. I did not understand why are they empty in this case.

What I actually would like to implement is :

I would like to make connection attempts to RMQ server - If the server is down. I would like my script to retry the connection until the server is up.Though I could not see a parameter wherein I could retry forever, I thought of trying it atleast for a predefined number of attempts. For this, I used connection_attempts and retry_delay connection parameters.

That resulted in the above select error when RMQ server is down.

Am I missing anything?

**Comment by vitaly-krugl** [comment]
Thanks for the info. It's a Windows-specific bug in pika. When there aren't any file descriptors, the code should be using time.sleep() (or something like that) instead of select.select(). It's a simple fix, but not a high-priority one. Your app can leave connection_attempts at 1, and have your own simple loop that performs the retries

Thanks for reporting.

**Comment by vitaly-krugl** [comment]
@GayatriNittala, this should now be fixed in pika's master branch as part of PR #735. See https://github.com/vitaly-krugl/pika/blob/c36b16e5aaaa82f938afa91a65432a4b82da2a56/pika/adapters/select_connection.py#L708-L713
